File: src/args.c
// SPDX-License-Identifier: GPL-3.0-only
/*
 *  sud/src/args.c
 *
 *  Copyright (C) Emily <info@emy.sh>
 */

#include <argp.h>
#include <errno.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <sud/args.h>
#include <sud/sud.h>
#include <sud/utils.h>
#include <sys/types.h>

#define DAEMON_OPTION  0x80
#define ISOLATE_OPTION 0x81
#define COLOR_OPTION   0x82

const char *argp_program_version = "0.1";
const char *argp_program_bug_address = "<erny@castellotti.net>";

static struct argp_option options[] = {
    {"color", COLOR_OPTION, "color", 0, "Set background color (default = \"41\")", 0},

    {"daemon", DAEMON_OPTION, 0, 0, "Start SUD as server (must be root)", 0},

    {"shell", 's', 0, 0, "Run shell as the target user", 0},

    {"user", 'u', "user", 0, "Run command as specified user name or ID", 0},

    {"non-interactive", 'n', 0, 0, "Non-interactive mode", 0},

    {"stdin", 'S', 0, 0, "Read password from standard input", 0},

    {"isolate", ISOLATE_OPTION, "policy", 0, "Apply sandboxing policies to the process", 0},

    {"version", 'V', 0, 0, "Display version information and exit", 0},

    {"help", 'h', 0, 0, "Display help message and exit", 0},

    {0}
};

static error_t parse_opt(int key, char *arg, struct argp_state *state) {
    int uid;
    int isolate;
    sud_cmdline_args_t *arguments = state->input;

    switch (key) {
        case COLOR_OPTION: {
            arguments->background_color = arg;
            break;
        }

        case DAEMON_OPTION: {
            if (state->argc > 2) {
                argp_error(state, "option '--daemon' cannot be used with other options");
                return ARGP_KEY_ERROR;
            }

            arguments->flags |= SUD_F_DAEMON;
            break;
        }

        case 's': {
            arguments->flags |= SUD_F_SHELL;
            break;
        }

        case 'u': {
            uid = get_uid_from_str(arg);
            if (uid < 0) {
                argp_error(state, "invalid uid");
                return ARGP_KEY_ERROR;
            }

            arguments->user = uid;

            break;
        }

        case 'n': {
            arguments->flags |= SUD_F_NOINT;
            break;
        }

        case 'S': {
            arguments->flags |= SUD_F_STDIN;
            break;
        }

        case ISOLATE_OPTION: {
            isolate = parse_isolate_str(arg);
            if (isolate < 0) {
                argp_error(state, "isolate option parse fail!");
                return ARGP_KEY_ERROR;
            }

            arguments->isolate |= isolate;
            break;
        }

        case ARGP_KEY_ARGS: {
            arguments->argc = state->argc - state->next;
            arguments->argv = state->argv + state->next;
            break;
        }

        case 'h': {
            printf("Super User Daemon - privilege manager for systemd/Linux\n");
            argp_state_help(state, state->out_stream, ARGP_HELP_STD_HELP);
            break;
        }

        case ARGP_KEY_FINI: {
            if ((arguments->flags & SUD_F_SHELL) && arguments->argc != 0) {
                argp_error(state, "option '--shell' expects no command arguments");
                return ARGP_KEY_ERROR;
            }

            break;
        }

        default: {
            return ARGP_ERR_UNKNOWN;
        }
    }

    return 0;
}

static struct argp argp = {options, parse_opt, "[command [arg ...]]", 0, 0, 0, 0};

int __parse_cmdline(int argc, char *argv[], sud_cmdline_args_t *args, int options) {
    memset(args, 0, sizeof(sud_cmdline_args_t));

    args->background_color = "41";

    return argp_parse(&argp, argc, argv, options, 0, args);
}

int parse_cmdline(int argc, char *argv[], sud_cmdline_args_t *args) {
    return __parse_cmdline(argc, argv, args, ARGP_IN_ORDER | ARGP_NO_HELP);
}

int parse_cmdline_silence(int argc, char *argv[], sud_cmdline_args_t *args) {
    return __parse_cmdline(argc, argv, args, ARGP_IN_ORDER | ARGP_SILENT);
}

File: src/args.rs
// SPDX-License-Identifier: GPL-3.0-only
/*
 *  sud/src/args.rs
 *
 *  Copyright (C) Emily <info@emy.sh>
 */

use crate::auth::UserInfo;
use crate::sud;
use crate::utils::{ProcessInfo, find_executable};
use clap::Parser;

#[derive(Debug, PartialEq, Clone)]
pub enum SudIsolateFlags {
    System,
    SystemFull,
    SystemStrict,
    Home,
    HomeRo,
    HomeTmpfs,
    Tmp,
    Devices,
    Net,
    User,
    Ktuneables,
    Klogs,
}

impl SudIsolateFlags {
    pub fn as_str(&self) -> &'static str {
        match self {
            SudIsolateFlags::System => "system",
            SudIsolateFlags::SystemFull => "system-full",
            SudIsolateFlags::SystemStrict => "system-strict",
            SudIsolateFlags::Home => "home",
            SudIsolateFlags::HomeRo => "home-ro",
            SudIsolateFlags::HomeTmpfs => "home-tmpfs",
            SudIsolateFlags::Tmp => "tmp",
            SudIsolateFlags::Devices => "devices",
            SudIsolateFlags::Net => "net",
            SudIsolateFlags::User => "user",
            SudIsolateFlags::Ktuneables => "kernel-tunables",
            SudIsolateFlags::Klogs => "kernel-logs",
        }
    }
}

#[derive(Parser, Debug, Default)]
#[command(about, version, before_help = "SUD - Super User Daemon")]
pub struct SudCmdlineArgs {
    #[arg(long, help = "Set background color")]
    pub color: Option<String>,

    #[arg(
        long,
        default_value = "false",
        exclusive = true,
        help = "Start SUD as server (must be root)"
    )]
    pub daemon: bool,

    #[arg(
        long,
        short,
        default_value = "false",
        conflicts_with = "command",
        help = "Run shell as the target user"
    )]
    pub shell: bool,

    #[arg(
        long,
        short,
        default_value = "0",
        help = "Run command as specified user name or ID"
    )]
    user: String,

    #[arg(
        long = "non-interactive",
        short = 'n',
        default_value = "false",
        help = "Non-interactive mode"
    )]
    pub non_interactive: bool,

    #[arg(
        long,
        short = 'S',
        default_value = "false",
        help = "Read password from standard input"
    )]
    pub stdin: bool,

    #[arg(long, help = "Run command in a specified working directory")]
    pub workdir: Option<String>,

    #[arg(long, num_args=1.., help="Apply sandboxing policies to the process")]
    isolate: Vec<String>,

    #[arg(
        long,
        default_value = "false",
        exclusive = true,
        help = "Clear persistent authentications of the current user"
    )]
    pub clear_persist: bool,

    #[arg(
        long,
        default_value = "false",
        exclusive = true,
        help = "Clear persistent authentications for all users (alias of --clear-persist if the user is not root)"
    )]
    pub clear_persist_all: bool,

    #[arg(num_args=1, trailing_var_arg=true, allow_hyphen_values=true, required_unless_present_any(["daemon", "shell", "clear_persist", "clear_persist_all"]))]
    pub command: Option<String>,

    #[arg(num_args=1..)]
    pub args: Vec<String>,
}

impl SudCmdlineArgs {
    pub fn parse(pinfo: &ProcessInfo) -> Result<Self, sud::SudError> {
        let mut args = SudCmdlineArgs::try_parse_from(&pinfo.argv)?;

        args.workdir = match args.workdir {
            Some(workdir) => Some(workdir),
            None => Some(pinfo.cwd.to_string_lossy().to_string()),
        };

        if args.shell {
            args.command = match pinfo.get_env("SHELL") {
                Some(env) => Some(env),
                None => Some(args.get_user()?.user.shell),
            };
        }

        if args.command.is_some() {
            let process_path = pinfo.get_env("PATH").ok_or(sud::SudError::NotFound(
                "Missing env PATH in caller process".into(),
            ))?;

            args.command = Some(
                find_executable(
                    &args.command.unwrap(),
                    &process_path,
                    &args.workdir.clone().unwrap(),
                )
                .ok_or(sud::SudError::NotFound("Command not found in PATH".into()))?,
            );
        }

        Ok(args)
    }

    pub fn get_isolate(&self, flag_check: SudIsolateFlags) -> bool {
        self.isolate.iter().any(|s| s == flag_check.as_str())
    }

    fn get_multi_isolate<const N: usize>(
        &self,
        flags: [SudIsolateFlags; N],
    ) -> Option<SudIsolateFlags> {
        for flag in flags {
            if self.get_isolate(flag.clone()) {
                return Some(flag);
            }
        }
        None
    }

    pub fn get_isolate_system(&self) -> Option<SudIsolateFlags> {
        self.get_multi_isolate([
            SudIsolateFlags::SystemStrict,
            SudIsolateFlags::SystemFull,
            SudIsolateFlags::System,
        ])
    }

    pub fn get_isolate_home(&self) -> Option<SudIsolateFlags> {
        self.get_multi_isolate([
            SudIsolateFlags::Home,
            SudIsolateFlags::HomeTmpfs,
            SudIsolateFlags::HomeRo,
        ])
    }

    pub fn get_user(&self) -> Result<UserInfo, sud::SudError> {
        UserInfo::from_str(&self.user)
    }
}
